### PR TITLE
ARROW-9501: [C++][Gandiva] Add logic in timestampdiff() when end date is last day of…

### PR DESCRIPTION
**Pull request by sgnkc**
… a month

timestampdiff(month, startDate, endDate) returns wrong result in Gandiva when the endDate < startDate and endDate is the last day of the month. An additional month is said to have passed when the end day is greater than or equal to the start day, but this does not hold true for dates which are last days of the month.

Case in point, if startDate = 2020-01-31, endDate = 2020-02-29, previously timestampdiff() returned 0, but the correct result should be 1.

### CODE REVIEW

File: cpp/src/gandiva/precompiled/timestamp_arithmetic.cc [hunk 1]
```diff
@@ -66,7 +96,7 @@ extern "C" {
     if (end_tm.TmMday() < start_tm.TmMday()) {                                        \
       /* case b */                                                                    \
       diff = MONTHS_TO_TIMEUNIT(months_diff - 1, N_MONTHS);                           \
-      return SIGN_ADJUST_DIFF(is_positive, diff);                                     \
+      return SIGN_ADJUST_DIFF(is_positive, diff) + is_last_day_of_month(end_tm);      \
```
Comment (vvellanki) [inline cpp/src/gandiva/precompiled/timestamp_arithmetic.cc:99]: Change this to:
is_last_day ? 1 : 0

Lets not add a bool to an int

**Review by praveenbingo — APPROVED** [review]

**Comment by praveenbingo** [comment]
thanks @sagnikc-dremio

**Comment by kszucs** [comment]
@praveenbingo please don't forget to add the relevant components `[C++][Gandiva]` to the pull request's title.

**Comment by kszucs** [comment]
I mean before squashing the PR to the master branch :)

**Comment by praveenbingo** [comment]
Sorry @kszucs missed that. Will monitor in future before merging.

**Pull request closed without merging**